src/main/io/gps.c

```diff
@@ -1805,22 +2352,19 @@ static bool gpsNewFrameUBLOX(uint8_t data)
             _step++;
             _ck_b += (_ck_a += data);       // checksum byte
             _payload_length += (uint16_t)(data << 8);
-#if DEBUG_UBLOX_FRAMES
-    debug[3] = _payload_length;
-#endif
-            if (_payload_length > UBLOX_PAYLOAD_SIZE) {
-                _skip_packet = true;
-            }
+
             _payload_counter = 0;   // prepare to receive payload
             if (_payload_length == 0) {
                 _step = 7;
             }
+            _buffer.bytes[_payload_counter] = data;
+            if (_payload_length > UBLOX_PAYLOAD_SIZE) {
```

Comment:
This is wrong. This will truncate received message, so checksum will fail

Reply:
It all depends on how we define "wrong" in this context.

Regarding increasing the `gpsData.errors` when we receive a message that exceeds our buffer's hard limit and is too long for us to process, it might be a good idea to do so (which is what currently happens).

However, when it comes to checking the checksum of a message that we definitely won't be able to process, it might not be necessary. It's best not to perform arithmetics that serve no purpose.

If you strongly believe that errors should not be incremented and that we should still process the checksum for messages we can't handle, we can add extra complication here.

But, honestly, I don't see much benefit in doing that, the message is already lost.

Reply:
Sat messages can easily exceed payload length; lets say the module sees 35 sats, while our internal max sats payload size is (and it currently is) 32 sat packets.

Reply:
If that happens, the message will not be processed anyway. Increasing the buff size is the right thing to do (if we expect this to really happen).
On the flip side, if let's say `FF FF` gets into the length fields due to random noise/read error, we *really* should not try to process the entire message and calculate the checksum.

Reply:
@adam-ah can you please suggest alternate code to handle @ledvinap's comments, otherwise perhaps I should revert the commit that included your changes.

Reply:
@ctzsnooze do whatever you think you should, it is still wrong to keep reading outside of our processing boundaries. Again, if an incorrect length gets in (an extreme case would be FF FF) you'd wipe out the following hundreds of messages (tens of seconds) if you don't stop at a hard boundary.
 If you revert the code, that's what you are signing up to do. Will it affect some users at some points? Probably. 

The only correct thing to do is to fix our buffer size to actually fit the messaged we expect to come, and if they don't fit then stop processing and start the next one. There is no good reason not to have a buffer that can accommodate the expected messages.

Reply:
No, that's not the "only correct thing to do."

We need to limit the overall buffer size due to Betaflight space limitations. We can't have a buffer that is as long as the 56 possible sats we might get. And no need for it, the only current use for sat data in BF is to return it to MSP clients (Configurator) so people can tell if sats are being acquired. If someone wants to analyze all possible sats the module is seeing, they can use a different tool for that.

We can have code that, when the message size isn't corrupted, will load the payload up to the max sats we want to support, skip the rest of the payload. and correctly calculate the checksum for the entire payload so we don't just toss the entire message needlessly. That's what the code was doing before it was broken.

We can also add a new check that would abort the message if we seem to be getting way too many bytes for a message, perhaps because the size bytes got corrupted or some other reason. We could do that right when we get a size value that seems suspicious, and not lose potential legit messages afterwards.

Reply:
Can you please help me to understand what space limitations we will run into if we allocate the correct buffer size, in the context of 32 is fine, 56 will break something?
Mind you, we just increased the UART buffer size from 128 to 256 without thinking much of it so it doesn't always seem to be an issue.
Not having the right buffer size here cause several problems (need extra check for checksum, need extra check for excessively long messages, need to cater for parsing truncated satellite messages)

Reply:
@adam-ah 
We did not increase the uart buffer size to 256 'without thinking about it'.  We logged the buffer usage and found that 256 was enough but 128 was not.  Logs show currently that at 115200 we hit peaks in the uart buffer around 233 but not above.  I showed you that data previously.

Configurator only displays data on up to 32 sats, and reports the number of sats in the solution (which we get from nav-pvt anyway.  So we can ignore data from more than 32 sats at this point, the user won't see it anyway, and there's no need to show more than 32.  The purpose of the display is to show their strength and stability, so that the user can angle the quad and check that none involved in the 3D fix drop out, which will cause significant glitching in the received data.  Showing info on 32 sats is more than adequate for that purpose.

Please reflect on your attitude; generally, it does you no credit.

Reply:
This is exactly what I meant by not thinking about it much. It was based on your measurement that showed a number that should not be possible given the maximal speed of the communication and the frequency of checking and reading the serial buffer.  

If you do the math, you'll see 233 given the above is not possible unless we have a conceptual problem somewhere. This 233 was reached without saturating the full 115200 bandwidth which is even more concerning.

Let's stick to addressing facts, trust me, no one would benefit from being commented on their attitude

Reply:
Reverted [42fc8c04f](https://github.com/betaflight/betaflight/pull/12799/commits/42fc8c04fdbc436c9ef196b88f0764ffcbb9239f)

Reply:
Coming back to this, a dynamically resized array would be the right solution, the code for that is actually rather simple. We could allocate the minimum actually required array (`MIN(NAV_DOP, NAV_PVT)=92`) which would be enough for most people most of the time. If someone enables `NAV_SAT` as well, we would increment dynamically the size of that array to some absolute maximum size (`8+12*56 sats`). In the unlikely event f we run into memory issues, it would only affect testing via configurator (otherwise we would not receive `NAV_SAT` messages).

This would eliminate the fragile and side-effect-prone problems discussed above.

I realise this is a bigger change and does not belong to this PR.

In any case, a change *is* required here: if the reading exceeds the absolute max that we should ever see (`8+12*56 `) then stop processing as it's due to an error, and we are unreasonably blocking the processing of good messages.

@ctzsnooze could you please add this (or something similar, depending how you would want to define constants) to step 5:

```
#define UBLOX_PAYLOAD_SIZE_ABSOLUTE_MAX (8 + 12 * 56)
[...]

            if (_payload_length == 0 || _payload_length >UBLOX_PAYLOAD_SIZE_ABSOLUTE_MAX) {
                _step = 7;
            }
```

This is a cheap check and allows us to resume finding the next preamble rather than wasting up to 64k of valuable data.

Reply:
> It all depends on how we define "wrong" in this context.
Wrong as 'sub optimally handles one problem and breaks existing code'

> * The estimation of absolute max message length was wrong (`UBLOX_PAYLOAD_SIZE` is not correct)
UBLOX_PAYLOAD_SIZE is intentionally smaller than maximum value possible

> * Data reading problems lead to the length field holding incorrect values
This is problem that should be handled. Cut correct reaction is to ignore packet immediately and start searching for next preamble. Changing size to  UBLOX_PAYLOAD_SIZE will mean that GPS is blocked for same time (UBLOX_PAYLOAD_SIZE), then checksum fails and synchronization search starts.

> In either case, there is either no point or utterly wrong to keep processing the message / checksum.
Only second case.

> More broadly, when it comes to calculating the checksum of a message that we definitely won't be able to process, it might not be necessary - no need to perform operations that serve no purpose.

We are perfectly able pro process only part of message that returns array of values. Code before this change did it.

> If you strongly believe that errors should not be incremented and that we should still process the checksum for messages we can't handle (for reasons above), we can add extra complication here.
> But honestly, I don't see much benefit in doing that, the message is already lost and/or wrong.

See above

Reply:
I agree with all of this @ledvinap apart from the missing comments / unit tests.
The current implementation badly needs at least source code comments to make it obvious that both the exact buffer size (8+N*12) and continuing calculating the checksum are critical.
Ideally a unit test would cover this, but we are a bit far from that, so let's settle on line comments to make this clear? The current implementation of step 6 reads more like a bug than a feature that supports parsing truncated NAV-SAT.

Otherwise see my comment above about adding an extra check for too long messages.

Reply:
> Coming back to this, a dynamically resized array would be the right solution, the code for that is actually rather simple. We could allocate the minimum actually required array (`MIN(NAV_DOP, NAV_PVT)=92`) which would be enough for most people most of the time. If someone enables `NAV_SAT` as well, we would increment dynamically the size of that array to some absolute maximum size (`8+12*56 sats`). In the unlikely event f we run into memory issues, it would only affect testing via configurator (otherwise we would not receive `NAV_SAT` messages).

Introducing memory management into BF is very problematic. It will cause more problems than it will fix,

> @ctzsnooze could you please add this (or something similar, depending how you would want to define constants) to step 5:
> ```
> #define UBLOX_PAYLOAD_SIZE_ABSOLUTE_MAX (8 + 12 * 56)
> [...]

             if (_payload_length == 0) {
                 _step = 7;
             }
             if (_payload_length > UBLOX_PAYLOAD_SIZE_ABSOLUTE_MAX) {
                 _skip_packet = true; 
                 _step = 7;
             }

> This is a cheap check and allows us to resume finding the next preamble rather than wasting up to 64k of valuable data.

Reply:
@ctzsnooze pls don't forget to add the check (above) to step 6